runtimes/d8.js

```diff
@@ -38,4 +38,109 @@ var $ = {
   IsHTMLDDA() { return {}; },
   source: $SOURCE,
   realm: Realm.current(),
+  agent: (function() {
+    /*
+      Copyright 2017 the V8 project authors. All rights reserved.
+      Use of this source code is governed by a BSD-style license that can be
+      found in the https://chromium.googlesource.com/v8/v8.git/+/master/LICENSE file.
+      This agent source is adapted from
+      https://chromium.googlesource.com/v8/v8.git/+/master/test/test262/harness-agent.js
+    */
+    var workers = [];
+    var i32a = null;
+    var pendingReports = [];
+
+    // Agents call Atomics.wait on this location to sleep.
+    var SLEEP_LOC = 0;
+    // 1 if the started worker is ready, 0 otherwise.
+    var START_LOC = 1;
+    // The number of workers that have received the broadcast.
+    var BROADCAST_LOC = 2;
+    // Each worker has a count of outstanding reports; worker N uses memory
+    // location [WORKER_REPORT_LOC + N].
+    var WORKER_REPORT_LOC = 3;
+
+    function workerScript(script) {
+      return `
+        var index;
+        var i32a = null;
+        var broadcasts = [];
+        var pendingReceiver = null;
+        function handleBroadcast() {
+          if (pendingReceiver && broadcasts.length > 0) {
+            pendingReceiver.apply(null, broadcasts.shift());
+            pendingReceiver = null;
+          }
+        };
+
+        onmessage = function(msg) {
+          switch (msg.kind) {
+            case 'start':
+              i32a = msg.i32a;
+              index = msg.index;
+              (0, eval)(\`${script}\`);
+              break;
+            case 'broadcast':
+              Atomics.add(i32a, ${BROADCAST_LOC}, 1);
+              broadcasts.push([msg.sab, msg.id]);
+              handleBroadcast();
+              break;
+          }
+        };
+        var $262 = {
```

Comment:
Looking at this skeptically...

Reply:
Here's an example of how this is being used in practice: https://github.com/tc39/test262/blob/ce9419779fc61d5cdfa37a5cda7b6828c67221da/test/built-ins/Atomics/wait/did-timeout.js#L13-L20